src/main.rs

```diff
@@ -0,0 +1,150 @@
+use crate::postgres::Postgres;
+use actyxos_sdk::event_service::{EventService, Subscription};
+use anyhow::Result;
+use backtrace::Backtrace;
+use db::{Db, DbConnection, DbEvent};
+use futures::{future::FutureExt, stream::StreamExt};
+use futures_batch::ChunksTimeoutStreamExt;
+use serde::{Deserialize, Serialize};
+use std::{env, time::Duration};
+use structopt::StructOpt;
+use tokio_compat_02::FutureExt as FutureExt02;
+use tracing::*;
+
+mod db;
+mod postgres;
+
+#[derive(StructOpt, Debug, Serialize, Deserialize)]
+#[structopt(
+    about = "Insert events into a Postgres database.\nEnvironment variables:\n  PGPASSWORD: PostgreSQL password, if not provided with --password/-p\n  AX_EVENT_SERVICE_URI: URL to connect to the Event Service. Default: http://localhost:4454/api/"
+)]
+struct Opt {
+    #[structopt(long, short)]
+    from_start: bool,
+    #[structopt(
+        long,
+        short = "w",
+        about = "If not provided, read from the PGPASSWORD environment variable"
+    )]
+    password: Option<String>,
+    #[structopt(long, short, default_value = "1024")]
+    max_batch_size: usize,
+    #[structopt(long, short)]
+    db_name: String,
+    #[structopt(long, short, default_value = "localhost")]
+    host: String,
+    #[structopt(long, short, default_value = "5432")]
+    port: u16,
+    #[structopt(long, short)]
+    username: String,
+    #[structopt(long, short, default_value = "events")]
+    table: String,
+    #[structopt(about = "Subscription set to listen to", default_value = "[{}]")]
+    subscriptions: String,
+}
+
+pub fn main() -> Result<()> {
+    env_logger::init();
+    init_panic_hook();
+    let opt: Opt = Opt::from_args();
+
+    let password = opt
+        .password
+        .unwrap_or_else(|| env::var("PGPASSWORD").expect("Postgres password must be provided"));
+
+    let subscriptions: Vec<Subscription> =
+        serde_json::from_str(opt.subscriptions.as_str()).unwrap();
+    info!("Subscribing to: {:?}", subscriptions);
+
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .worker_threads(2)
+        .enable_all()
+        .build()
+        .unwrap();
+
+    let pg = Postgres::new(
+        opt.host,
+        opt.port,
+        opt.username,
+        password,
+        opt.db_name,
+        opt.table,
+    );
+
+    runtime.block_on(run_pipeline(
+        Box::new(pg),
+        subscriptions,
+        opt.max_batch_size,
+    ))
+}
+
+async fn run_pipeline<C: DbConnection + 'static>(
+    db: Box<dyn Db<C>>,
+    subscriptions: Vec<Subscription>,
+    max_batch_size: usize,
+) -> Result<()> {
+    let event_service = EventService::default();
+    debug!("Connected to EventService");
+
+    let present = event_service.get_offsets().compat().await?;
```

Comment:
you could just put the `compat` inside your `main`, and not worry about it here anymore. Using `tokio_02_compat::FutureExt::compat` will enable the future to use both tokio 02 and 03 features, so no need to be frugal

Reply:
nice!